pontoon/sync/formats/json_keyvalue.py

```diff
@@ -0,0 +1,184 @@
+                    entity.order, entity.key, entity.context, "", None
+                )
+
+        try:
+            with codecs.open(path, "r", "utf-8") as resource:
+                self.json_file = json.load(resource)
+                validate(self.json_file, SCHEMA)
+
+        except (OSError, ValueError, ValidationError) as err:
+            # If the file doesn't exist or cannot be decoded,
+            # but we have a source resource,
+            # we can keep going, we'll just not have any translations.
+            if source_resource:
+                return
+            else:
+                raise ParseError(err)
+
+        self.order_count = 0
+
+        # Callback Populate JSON Entities
+        def readEntity(internal_key, dot_key, value):
+            self.entities[internal_key] = JSONEntity(
+                self.order_count, internal_key, dot_key, value, value
+            )
+            self.order_count += 1
+
+        self.traverse_json(self.json_file, readEntity)
+
+    @property
+    def translations(self):
+        return sorted(self.entities.values(), key=lambda e: e.order)
+
+    def save(self, locale):
+        """
+        Load the source resource, modify it with changes made to this
+        Resource instance, and save it over the locale-specific
+        resource.
+        """
+        if not self.source_resource:
+            raise SyncError(
+                "Cannot save JSON resource {}: No source resource given.".format(
+                    self.path
+                )
+            )
+
+        with codecs.open(self.source_resource.path, "r", "utf-8") as resource:
+            json_file = json.load(resource, object_pairs_hook=OrderedDict)
+
+        def writeEntity(internal_key, dot_key, value):
+            entity = self.entities[internal_key]
+            if entity.strings:
+                self.set_json_value(json_file, internal_key, entity.strings[None])
+            else:
+                self.del_json_value(json_file, internal_key)
+
+        self.traverse_json(json_file.copy(), writeEntity)
+        self.clear_empty_objects(json_file)
+
+        create_parent_directory(self.path)
+
+        with codecs.open(self.path, "w+", "utf-8") as f:
+            log.debug("Saving file: %s", self.path)
+            f.write(
+                json.dumps(
+                    json_file, ensure_ascii=False, indent=2, separators=(",", ": ")
+                )
+            )
+            f.write("\n")  # Add newline
+
+    # Recursively read json object
+    # Callback a function when reaching the end of a branch
+    def traverse_json(self, data, function, keys=[]):
+        for key, value in data.copy().items():
+            currentKey = keys.copy()
+            currentKey.append(key)
+            if isinstance(value, dict):
+                self.traverse_json(value, function, keys=currentKey)
+            elif type(value) == str:
+                internal_key = "<dot>".join(currentKey)
```

Comment:
Rather than using a custom format like this, why not use a string list as the `internal_key`? So e.g. the key in `{ "foo": { "bar": "msg" } }` would get encoded as `['foo', 'bar']`, while the key in `{ "foo.bar": "msg" } }` would be `['foo.bar']`.

Reply:
I've implemented this solution using JSON.dumps and loads!